emissary-core/src/primitives/mapping.rs

```diff
@@ -18,102 +18,127 @@
-        write!(f, "{}={}", self.key, self.value)
+        self.0.fmt(f)
     }
 }
 
 impl Mapping {
-    /// Create new [`Mapping`].
-    pub fn new(key: Str, value: Str) -> Self {
-        Self { key, value }
-    }
-
     /// Serialize [`Mapping`] into a byte vector.
-    pub fn serialize(self) -> Vec<u8> {
-        let key = self.key.serialize();
-        let value = self.value.serialize();
-
-        // key length + value length + length field + `=` + `;`
-        let size = key.len() + value.len() + 2;
-        let mut out = vec![0u8; size];
-
-        out[..key.len()].copy_from_slice(&key);
-        out[key.len()] = b'=';
-        out[1 + key.len()..1 + key.len() + value.len()].copy_from_slice(&value);
-        out[1 + key.len() + value.len()] = b';';
+    pub fn serialize(&self) -> Vec<u8> {
+        // Allocate at least two bytes for the size prefix
+        let mut out = BytesMut::with_capacity(2);
+        let mut data = out.split_off(2);
+
+        let mut entries: Vec<_> = self.0.iter().collect();
+
+        // Our mapping implementation does not support duplicate keys, so we do not need to preserve
+        // order
+        entries.sort_unstable_by(|a, b| a.0.cmp(b.0));
+        for (key, value) in entries {
+            let key = key.serialize();
+            let value = value.serialize();
+            data.reserve(key.len() + value.len() + 2);
+            data.extend(key);
+            data.put_u8(b'=');
+            data.extend(value);
+            data.put_u8(b';');
+        }
+        //TODO: validate that this is less than u16::MAX?
+        out.put_u16(data.len() as u16);
+        out.unsplit(data);
 
-        out
+        //TODO: Work with `Bytes` instead of `Vec<u8>` to avoid copying data?
+        out.freeze().to_vec()
     }
 
-    /// Parse [`Mapping`] from `input`, returning rest of `input` and parsed address.
-    pub fn parse_frame(input: &[u8]) -> IResult<&[u8], Mapping> {
-        let (rest, key) = Str::parse_frame(input)?;
-        let (rest, _) = be_u8(rest)?; // ignore `=`
-        let (rest, value) = Str::parse_frame(rest)?;
-        let (rest, _) = be_u8(rest)?; // ignore `;`
-
-        Ok((rest, Mapping { key, value }))
-    }
+    /// Parse [`Mapping`] from `input`, returning rest of `input` and parsed mapping.
+    pub fn parse_frame(input: &[u8]) -> IResult<&[u8], Self> {
+        let (rest, size) = be_u16(input)?;
+        let mut mapping = Self::default();
+
+        if let Some((mut data, rest)) = rest.split_at_checked(size as usize) {
+            while !data.is_empty() {
+                let (remaining, key) = Str::parse_frame(data)?;
+                let (remaining, _) = be_u8(remaining)?; // Should we validate '='?
+                let (remaining, value) = Str::parse_frame(remaining)?;
+                let (remaining, _) = be_u8(remaining)?; // Should we validate ';'?
+                mapping.insert(key, value);
+                data = remaining;
```

Comment:
```suggestion
                let (remaining, key) = Str::parse_frame(data)?;
                let (remaining, _) = be_u8(remaining)?;
                let (remaining, value) = Str::parse_frame(remaining)?;
                let (remaining, _) = be_u8(remaining)?;

                mapping.insert(key, value);
                data = remaining;
```

I don't think we need to validate them